optuna/study.py

```diff
@@ -510,13 +510,16 @@ def _run_trial(self, func, catch):
             self.logger.info(message)
             self._storage.set_trial_state(trial_id, structs.TrialState.PRUNED)
             return trial
-        except catch as e:
+        except Exception as e:
             message = 'Setting status of trial#{} as {} because of the following error: {}'\
                 .format(trial_number, structs.TrialState.FAIL, repr(e))
             self.logger.warning(message, exc_info=True)
             self._storage.set_trial_system_attr(trial_id, 'fail_reason', message)
             self._storage.set_trial_state(trial_id, structs.TrialState.FAIL)
-            return trial
+
+            if any(isinstance(e, cls) for cls in catch):
+                return trial
+            raise
```

Comment:
We should not do a `raise e` as this will clear the callstack in Python 2. This `raise` won't.